Use Array.convert instead of Array.from, add compat layer for Array.from

SergioCrisostomo opened the pull request:
ES6 implemented `Array.from`, a method name that MooTools has since at least 2010. This pull request changes the MooTools implementation of `Array.from` to use the new native implementation. For older browsers that need a polyfill I added [MDN's polyfill](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/from) to make it available to older browsers also.

For compatibility and consistency added `Array.convert` which is the old `Array.from` implementation, thus making it easier to update pre-change code.
The old `Array.from` implementation is not changed in the _compat layer_, still overriding the native implementation.

Since the old implementation was overwriting the native I think we don't need to change older versions.

Tested everything in Travis and got green.

Fixes #2757

anutron commented:
I still think we should alias all the other `.from` static methods on the other natives to use `.convert` and make `.from` deprecated (but still present) for everything but Array.

/my 2 cents

Code LGTM.

arian commented on Source/Core/Core.js:
```diff
@@ -278,7 +278,7 @@ force('String', String, [
 	'slice', 'split', 'substr', 'substring', 'trim', 'toLowerCase', 'toUpperCase'
 ])('Array', Array, [
-	'pop', 'push', 'reverse', 'shift', 'sort', 'splice', 'unshift', 'concat', 'join', 'slice',
+	'pop', 'push', 'reverse', 'shift', 'sort', 'splice', 'unshift', 'concat', 'join', 'slice', 'from',
```
Aren't those only for the Array.prototype?

arian commented on Source/Core/Core.js:
```diff
@@ -294,9 +294,76 @@ force('String', String, [
+})());
+/*</!ES6>*/
+
+/*<1.6compat>*/
```
That should be 1.5 compat right?

arian commented:
Should we add the Array.from? I think just changing Array.from to Array.convert should do the trick. Historically `$splat` or `Array.from` is used internally, because it's useful for other things, but also to be used outside. Inside MooTools there isn't really a use case for the new `Array.from`.

Additionally I suspect people that want to use the new `Array.from` either assume a modern environment, or use `babel/polyfill` or similar.

arian commented:
Maybe you should do a search-replace through the entire docs, as `Array.from` is mentioned here too: http://mootools.net/core/docs/1.5.2/Core/Core#Deprecated-Functions:splat

arian commented:
If we decide to add the polyfill, we also need to test it ;)

anutron commented:
@arian  I think we should maintain backwards compat for users who need it whenever possible. I.e. if the user downloads the compat version Array.from is there. I'm not sure if that's what you're suggesting when you ask if we should add it. Looking at my own work, there are 11 instances in all my Behavior stuff. I can search and replace it to upgrade, that's fine, but my point is that it's not only used in Core.

SergioCrisostomo commented:
I agree with renaming, deprecating it and keep it only in the compat layer. I think that is what @arian also meant. 

If so I can update it, and take into account the things @arian pointed out in review.

arian commented:
@anutron of course we should maintain backwards compat (with the compat layer stuff), so `Array.from = Array.convert`.

What I'm not sure of is whether we should add the *new* ES6 version as polyfill.

anutron commented:
Ah. I gotcha.

How complete is our ES6 polyfill currently? Is `Array.from` like, the only thing missing? Do we even profess to offer an ES6 shim? Those are easily found in stand alone projects, right? I think I'd vote no, unless this is the only thing we don't fill in.

arian commented:
Looking at this pull request, @SergioCrisostomo just copied the one from MDN, which is likely pretty complete. However I don't think it's a good idea to just include it, as there are many other polyfills (or modern browsers) that developers can easily add themselves if needed.

Of course, as I said earlier, if MooTools itself would make heavy use of this functionality, such as other ES5 methods like bind/forEach/map, then it would make more sense.

SergioCrisostomo commented:
@arian @anutron agree that polyfill is very easy to find if needed. So I updated the PR, without it and fixing the things @arian pointed out.

SergioCrisostomo commented:
Updated the pull request as sugested in discussion and fixing the things arian pointed out.

@arian @anutron I agree that polyfill is very easy to find if needed, so I removed the polyfill from the PR.

SergioCrisostomo commented:
bump... :)
merge anyone?

anutron commented:
LGTM. Ship it.

anutron commented:
Sergio, why don't you go ahead and ship this to master? I don't think you need someone else to do it after the review is done...

arian commented on Docs/Types/Array.md:
```diff
@@ -778,11 +774,28 @@ Converts an RGB color value to hexadecimal. Input array must be in one of the fo
+
+This method has been deprecated in MooTools 1.6, please use *[Array:convert][]* instead. 
+For backwards compatibility you can use the _compat layer_ that still uses the old implementation, overriding the Native ES6 implementation. 
+Please use instead _no compat_ version, unless you know why you have to use the _compat layer_.
```
Doesn't read nicely?

> Please use the _no compat_ version instead, unless ...

arian commented on Source/Core/Core.js:
```diff
@@ -118,6 +118,11 @@ Function.prototype.implement = function(key, value){
 	this.prototype[key] = value;
 }.overloadSetter();
 
+Array.convert = function(item){
```
Nicer to move *below* the `slice` declaration.

arian commented:
One major concern, however, is the API consistency with `String.from` / `Number.from` / `Function.from`.

I suppose all of those need to change too, otherwise it would be inconsistent and weird.

anutron commented:
I made that comment earlier. Don't disagree. I would only remove Array.from though; you could leave the old ones there and alias it to both .from and .convert, then just update the docs to refer to .convert.

arian commented:
+1 to *convert* `.convert` for those too, like you said.

SergioCrisostomo commented:
@arian I am ok with removing/renaming them also. 

Can I do that in a separate PR?

arian commented:
:+1:
